exporter/prometheus.go

```diff
@@ -0,0 +1,454 @@
+package exporter
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/cadvisor/info"
+	"github.com/google/cadvisor/manager"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const namespace = "container"
+
+// Just a small interface for easier housekeeping
+type vector interface {
+	Reset()
+	Describe(ch chan<- *prometheus.Desc)
+	Collect(ch chan<- prometheus.Metric)
+}
+
+type PrometheusCollector struct {
+	mutex   sync.RWMutex
+	manager manager.Manager
+
+	errors   prometheus.Counter
+	lastSeen *prometheus.GaugeVec
+
+	cpuUsageSeconds       *prometheus.CounterVec
+	cpuUsageSecondsPerCPU *prometheus.CounterVec
+
+	memoryUsageBytes *prometheus.GaugeVec
+	memoryWorkingSet *prometheus.GaugeVec
+	memoryFailures   *prometheus.CounterVec
+
+	fsLimit        *prometheus.GaugeVec
+	fsUsage        *prometheus.GaugeVec
+	fsReads        *prometheus.CounterVec
+	fsReadsSectors *prometheus.CounterVec
+	fsReadsMerged  *prometheus.CounterVec
+	fsReadTime     *prometheus.CounterVec
+
+	fsWrites        *prometheus.CounterVec
+	fsWritesSectors *prometheus.CounterVec
+	fsWritesMerged  *prometheus.CounterVec
+	fsWriteTime     *prometheus.CounterVec
+
+	fsIoInProgress *prometheus.GaugeVec
+	fsIoTime       *prometheus.CounterVec
+
+	fsWeightedIoTime *prometheus.CounterVec
+
+	networkRxBytes   *prometheus.CounterVec
+	networkRxPackets *prometheus.CounterVec
+	networkRxErrors  *prometheus.CounterVec
+	networkRxDropped *prometheus.CounterVec
+	networkTxBytes   *prometheus.CounterVec
+	networkTxPackets *prometheus.CounterVec
+	networkTxErrors  *prometheus.CounterVec
+	networkTxDropped *prometheus.CounterVec
+
+	tasksSleeping        *prometheus.GaugeVec
+	tasksRunning         *prometheus.GaugeVec
+	tasksStopped         *prometheus.GaugeVec
+	tasksUninterruptible *prometheus.GaugeVec
+	tasksIoWait          *prometheus.GaugeVec
+
+	vectors []vector
+}
+
+// New returns a new PrometheusCollector
+func NewPrometheusCollector(manager manager.Manager) *PrometheusCollector {
+	c := &PrometheusCollector{
+		manager: manager,
+		errors: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "errors_total",
+			Help:      "Errors while exporting container metrics.",
+		}),
+		lastSeen: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "last_seen",
+			Help:      "Last time a container was seen by the exporter",
+		},
+			[]string{"name", "id", "image"},
+		),
+		cpuUsageSeconds: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "cpu_usage_seconds_total",
+			Help:      "Total seconds of cpu time consumed.",
+		},
+			[]string{"name", "id", "image", "type"},
```

Comment:
It's more useful to expose CPU usage total not broken down since it is significantly more accurate than broken down. We can either have type=total,user,system or have a separate total and breakdown metric. I'm guessing the first is better? If we go that route we can rename to cpu_usage_seconds